drivers/crypto/crypto_stm32_hash.c

```diff
@@ -0,0 +1,186 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/crypto/crypto.h>
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/reset.h>
+#include <soc.h>
+
+#include "crypto_stm32_hash_priv.h"
+
+#define LOG_LEVEL CONFIG_CRYPTO_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(crypto_stm32_hash);
+
+#define DT_DRV_COMPAT st_stm32_hash
+
+static struct crypto_stm32_hash_session stm32_hash_sessions[CONFIG_CRYPTO_STM32_HASH_MAX_SESSIONS];
+
+static int crypto_stm32_hash_get_unused_session_index(const struct device *dev)
+{
+	struct crypto_stm32_hash_data *data = CRYPTO_STM32_HASH_DATA(dev);
+
+	k_sem_take(&data->session_sem, K_FOREVER);
+
+	for (int i = 0; i < CONFIG_CRYPTO_STM32_HASH_MAX_SESSIONS; i++) {
+		if (!stm32_hash_sessions[i].in_use) {
+			stm32_hash_sessions[i].in_use = true;
+			k_sem_give(&data->session_sem);
+			return i;
+		}
+	}
+
+	k_sem_give(&data->session_sem);
+	return -1;
+}
+
+static int stm32_hash_handler(struct hash_ctx *ctx, struct hash_pkt *pkt, bool finish)
+{
+	const struct device *dev = ctx->device;
+	struct crypto_stm32_hash_data *data = CRYPTO_STM32_HASH_DATA(dev);
+	struct crypto_stm32_hash_session *session = CRYPTO_STM32_HASH_SESSN(ctx);
+	HAL_StatusTypeDef status;
+
+	if (!pkt || !pkt->in_buf || !pkt->out_buf) {
+		LOG_ERR("Invalid packet buffers");
+		return -EINVAL;
+	}
+
+	if (!finish) {
+		LOG_ERR("Multipart hashing not supported yet");
+		return -ENOTSUP;
+	}
+
+	k_sem_take(&data->device_sem, K_FOREVER);
+
+	switch (session->algo) {
+	case CRYPTO_HASH_ALGO_SHA224:
+		status = HAL_HASHEx_SHA224_Start(&data->hhash, pkt->in_buf, pkt->in_len,
+						 pkt->out_buf, HAL_MAX_DELAY);
+		break;
+	case CRYPTO_HASH_ALGO_SHA256:
+		status = HAL_HASHEx_SHA256_Start(&data->hhash, pkt->in_buf, pkt->in_len,
+						 pkt->out_buf, HAL_MAX_DELAY);
+		break;
+	default:
+		k_sem_give(&data->device_sem);
+		LOG_ERR("Unsupported algorithm in handler: %d", session->algo);
+		return -EINVAL;
+	}
+
+	k_sem_give(&data->device_sem);
+
+	if (status != HAL_OK) {
+		LOG_ERR("HAL HASH computation failed (status=%d)", status);
+		return -EIO;
+	}
+
+	LOG_DBG("Hash computation successful");
+	return 0;
+}
+
+static int stm32_hash_begin_session(const struct device *dev, struct hash_ctx *ctx,
+				    enum hash_algo algo)
+{
+	int ctx_idx;
+	struct crypto_stm32_hash_session *session;
+
+	switch (algo) {
+	case CRYPTO_HASH_ALGO_SHA224:
+	case CRYPTO_HASH_ALGO_SHA256:
+		break;
+	default:
+		LOG_ERR("Unsupported hash algorithm: %d", algo);
+		return -EINVAL;
+	}
+
+	ctx_idx = crypto_stm32_hash_get_unused_session_index(dev);
+	if (ctx_idx < 0) {
+		LOG_ERR("No free session for now");
+		return -ENOSPC;
+	}
+
+	session = &stm32_hash_sessions[ctx_idx];
+	memset(&session->config, 0, sizeof(session->config));
+	memset(session->digest, 0, sizeof(session->digest));
```

Comment:
Sorry if I'm not understanding, but I could not find where these two are used in the driver. Could you please elaborate where those are used?